metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -63,90 +66,41 @@ type enricher struct {
 	watcher            kubernetes.Watcher
 	watcherStarted     bool
 	watcherStartedLock sync.Mutex
+	namespaceWatcher   kubernetes.Watcher
+	nodeWatcher        kubernetes.Watcher
 	isPod              bool
 }
 
 const selector = "kubernetes"
 
-// GetWatcher initializes a kubernetes watcher with the given
-// scope (node or cluster), and resource type
-func GetWatcher(base mb.BaseMetricSet, resource kubernetes.Resource, nodeScope bool) (kubernetes.Watcher, error) {
-	return GetNamedWatcher("", base, resource, nodeScope)
-}
-
-func GetNamedWatcher(name string, base mb.BaseMetricSet, resource kubernetes.Resource, nodeScope bool) (kubernetes.Watcher, error) {
-	config := kubernetesConfig{
-		AddMetadata: true,
-		SyncPeriod:  time.Minute * 10,
-	}
-	if err := base.Module().UnpackConfig(&config); err != nil {
-		return nil, err
-	}
-
-	// Return nil if metadata enriching is disabled:
-	if !config.AddMetadata {
-		return nil, nil
-	}
-
-	client, err := kubernetes.GetKubernetesClient(config.KubeConfig, config.KubeClientOptions)
-	if err != nil {
-		return nil, err
-	}
-
-	options := kubernetes.WatchOptions{
-		SyncTimeout: config.SyncPeriod,
-	}
-
-	log := logp.NewLogger(selector)
-
-	// Watch objects in the node only
-	if nodeScope {
-		nd := &kubernetes.DiscoverKubernetesNodeParams{
-			ConfigHost:  config.Host,
-			Client:      client,
-			IsInCluster: kubernetes.IsInCluster(config.KubeConfig),
-			HostUtils:   &kubernetes.DefaultDiscoveryUtils{},
-		}
-		options.Node, err = kubernetes.DiscoverKubernetesNode(log, nd)
-		if err != nil {
-			return nil, fmt.Errorf("couldn't discover kubernetes node: %w", err)
-		}
-	}
-
-	log.Debugf("Initializing a new Kubernetes watcher using host: %v", config.Host)
-
-	return kubernetes.NewNamedWatcher(name, client, resource, options, nil)
-}
-
 // NewResourceMetadataEnricher returns an Enricher configured for kubernetes resource events
 func NewResourceMetadataEnricher(
 	base mb.BaseMetricSet,
 	res kubernetes.Resource,
 	nodeScope bool) Enricher {
 
-	watcher, err := GetNamedWatcher("resource_metadata_enricher", base, res, nodeScope)
-	if err != nil {
-		logp.Err("Error initializing Kubernetes metadata enricher: %s", err)
-		return &nilEnricher{}
-	}
-
-	if watcher == nil {
+	config := validatedConfig(base)
+	if config == nil {
 		logp.Info("Kubernetes metricset enriching is disabled")
 		return &nilEnricher{}
 	}
 
-	metaConfig := metadata.Config{}
-	if err := base.Module().UnpackConfig(&metaConfig); err != nil {
+	watcher, nodeWatcher, namespaceWatcher := getPodMetadataWatchers(config, res, nodeScope)
```

Comment:
Is this function only for Pod Metadata Watchers? The `res` is not hardcoded to Pod and can have other values. So I suggest that the name of the function should be more generic, like `getResourceMetadataWatchers`

Reply:
done https://github.com/elastic/beats/pull/29133/commits/ec17b227c511988c8648073783dd2b3640515481